### ISSUE TITLE
RFC: Changing the behavior of the `external` builtin

### ISSUE THREAD

**Issue by krader1961** [elves/elvish#1135]
TL;DR:

I'm inclined to think `external` should wrap the output of `search-external` in its "external" data type.

The long version:

I was working on PR #1121 to improve coverage of the `external` builtin. This caused me to notice that it does not actually validate that the external command even exists. Nor that it has execute permission even if there is a file in `$E:PATH` with that name. Also, the object it returns contains the original command string; not the full path to the command (assuming it was found in `$E:PATH`).

The [`external` documentation](https://elv.sh/ref/builtin.html#external) is silent on all of these points. It does, however, imply that `x = (external man)` will result in `$x` invoking that command. That will fail in these cases:

a) There is no "man" command in `$E:PATH` at the time `external man` is run.

b) There was a "man" command in `$E:PATH` at the time `external man` was run but a subsequent modification of `$E:PATH` removed it from the search path.

Please note that I am not talking about changes to the state of the file system after `external` is executed. I recognize that `rm /path/to/a/command` will cause any previous path resolution of that command to be invalidated. This issue is solely about whether the result of `external` should do **any validation** of its argument. Also, whether the result should be the original command string or the absolute path to the command (assuming it exists and has execute permission).

I'm inclined to think `external` should wrap the output of `search-external` in its "external" data type.

P.S., There is no practical way to verify an external file is a valid program (at least in any portable fashion). The best `external` can do is check that it is found in `$E:PATH` and has execute permission. Basically what `search-external` does.

**Comment by xiaq** [comment]
This behavior is intentional so that `(external name)` always matches `$e:name~`.

However, it makes sense to add a `&search` option to `external`, which

* Does the searching and throws an exception if the external command is not found on PATH

* Stores the absolute path in the `ExternalCmd`.

**Comment by krader1961** [comment]
> This behavior is intentional so that (external name) always matches $e:name~ ....

That sounds like a reasonable approach. Anyone who wants to use the object returned by `external` as if the embedded command was used literally can continue to do so. Anyone wanting to ensure the resolution to an external program occurs when `external` is run, to guard against future changes to `$E:PATH` and that the program exists when `external` is run, can use the `&search` option.

Modifying the implementation to add a `&search` option is easy. The hard part is going to be modifying the `external` documentation.
